test/cast/discriminator_test.exs

```diff
@@ -0,0 +1,215 @@
+defmodule OpenApiSpex.CastDiscriminatorTest do
+  use ExUnit.Case
+  alias OpenApiSpex.{Cast, Schema}
+  alias OpenApiSpex.Cast.{Discriminator, Error}
+
+  # The discriminator we'll be using across the tests. Animal type was
+  # specifically chosen to help the examples be more clear since abstract
+  # objects can be difficult to grok in this context.
+  @discriminator "animal_type"
+
+  defp cast(ctx) do
+    Discriminator.cast(struct(Cast, ctx))
+  end
+
+  defp cast_cast(ctx) do
+    Cast.cast(struct(Cast, ctx))
+  end
+
+  # Maps an arbitrary string value to a schema
+  def build_discriminator_mapping(name, schema) do
+    %{name => schema_ref(schema)}
+  end
+
+  def schema_ref(%{title: title}), do: "#/components/schemas/#{title}"
+
+  def build_schema(title, properties) do
+    %Schema{
+      type: :object,
+      title: title,
+      properties: properties
+    }
+  end
+
+  def build_discriminator_schema(schemas, composite_keyword, property_name, mapppings \\ nil) do
+    %Schema{
+      type: :object,
+      discriminator: %{
+        propertyName: "#{property_name}",
+        mapping: mapppings
+      }
+    }
+    |> Map.put(composite_keyword, schemas)
```

Comment:
Not really an issue here, because this is just a test, but if this was non-test code, I would suggest this be written as:

```elixir
|> struct([{composite_keyword, schemas}])
```

Or:
```elixir
|> struct(%{composite_keyword => schemas})
```

Or if either of those are too weird to read, this works too:

```elixir
|> Map.replace!(composite_keyword, schemas)
```

(But that breaks the rule "never use `Map` functions on a struct".)

All three examples provide protection against adding a key to the struct that isn't part of the struct's definition.